mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -882,6 +882,132 @@ struct ForwardOpPattern : public ConvertOpToLLVMPattern<ForwardOp> {
     }
 };
 
+struct ReverseOpPattern : public ConvertOpToLLVMPattern<ReverseOp> {
+    using ConvertOpToLLVMPattern::ConvertOpToLLVMPattern;
+
+    LogicalResult match(ReverseOp op) const override { return success(); }
+
+    void rewrite(ReverseOp op, OpAdaptor adaptor,
+                 ConversionPatternRewriter &rewriter) const override
+    {
+
+        auto argc = op.getArgc();
+        auto resc = op.getResc();
+        SmallVector<Value> inputs;
+        SmallVector<Value> differentials;
+        SmallVector<Value> outputs;
+        SmallVector<Value> cotangents;
+        SmallVector<Value> residuals;
+        auto params = op.getArguments();
+
+        for (auto i = 0; i < argc * 2; i++) {
+            bool isDup = (i % 2) != 0;
+            Value val = params[i];
+            isDup ? differentials.push_back(val) : inputs.push_back(val);
+        }
+
+        auto upperLimit = (argc * 2) + (resc * 2);
+        for (auto i = argc * 2; i < upperLimit; i++) {
+            bool isDup = (i % 2) != 0;
+            Value val = params[i];
+            isDup ? cotangents.push_back(val) : outputs.push_back(val);
+        }
+
+        auto tapeCount = op.getTape();
+        auto uppestLimit = upperLimit + tapeCount;
+        for (auto i = upperLimit; i < uppestLimit; i++) {
+            residuals.push_back(params[i]);
+        }
+
+        SmallVector<Type> newFuncInputTys;
+
+        for (auto [in, diff] : llvm::zip(inputs, differentials)) {
+            newFuncInputTys.push_back(in.getType());
+            newFuncInputTys.push_back(diff.getType());
+        }
+
+        for (auto [out, cotan] : llvm::zip(outputs, cotangents)) {
+            newFuncInputTys.push_back(out.getType());
+            newFuncInputTys.push_back(cotan.getType());
+        }
+
+        SmallVector<Type> residualStructs;
+        auto converter = getTypeConverter();
+        for (auto residualMemref : residuals) {
+            auto residualMemrefTy = residualMemref.getType();
+            auto residualStructTy = converter->convertType(residualMemrefTy);
+            residualStructs.push_back(residualStructTy);
+        }
+
+        auto ctx = rewriter.getContext();
+        // This gives me a struct of the following type
+        // { memref0, memref1, memref2, ... memrefN }
+        auto tapeTy = LLVM::LLVMStructType::getLiteral(ctx, residualStructs);
+        // This gives me { { memref, ..., memrefn } }
+        auto wrappedFlatTapeStructTy = LLVM::LLVMStructType::getLiteral(ctx, {tapeTy});
+        auto ptrTy = LLVM::LLVMPointerType::get(ctx);
+        Type retty = tapeCount > 0 ? dyn_cast<Type>(wrappedFlatTapeStructTy) : dyn_cast<Type>(ptrTy);
```

Comment:
Shouldn't call it retty

Reply:
https://github.com/PennyLaneAI/catalyst/pull/822/commits/8411b6228c7bbbb46336d15f12e3336939ded337